### PR TITLE
Refacto function assemble librz/asm/arch/z80/z80asm.c

### PR DESCRIPTION
**Pull request by ghost**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Refacto function assemble because of a dead do while

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/asm/arch/z80/z80asm.c [hunk 1]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
+						break;
+					}
+					wrtb (0x02 + 0x10 * (r == ld_DE));
 					break;
-				}
-				if (r == A_NN) {
-					wrtb (0x3A);
+				case ld_HL:
+					r = rd_ld_hl (&ptr);
+					wrtb (0x70 + --r);
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: for readness need, can you split the line in two? `r--` in its line and `0x70 + r` in wrtb ?

File: librz/asm/arch/z80/z80asm.c [hunk 2]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
 					break;
-				}
-				if (r == A_I || r == A_R) {
+				case ld_NN:
+					if (!(r = rd_ld_nn (&ptr))) {
+						break;
+					}
+					if (r == ld_nnA || r == ld_nnHL) {
+						wrtb (0x22 + 0x10 * (r == ld_nnA));
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: please avoid doing multiplication with compare. split this or do `(r == ld_nnA ? 1 : 0)`

File: librz/asm/arch/z80/z80asm.c [hunk 3]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
+					if (!(r = rd_lda (&ptr))) {
+						break;
+					}
+					if (r == A_NN) {
+						wrtb (0x3A);
+						break;
+					}
+					if (r == A_I || r == A_R) {
+						wrtb (0xED);
+						wrtb (0x57 + 8 * (r == A_R));
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: same as above: `(r == A_R ? 1 : 0)`

File: librz/asm/arch/z80/z80asm.c [hunk 4]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
+						break;
+					}
+					if (r == A_N) {
+						char n = rz_num_math (NULL, readbyte);
+						wrtb (0x3E);
+						wrtb (n);
+						break;
+					}
+					if (r < 0) {
+						wrtb (0x0A - 0x10 * ++r);
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: same as above for r++ and wrtb..

File: librz/asm/arch/z80/z80asm.c [hunk 5]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
-				s = rd_nn_nn (&ptr);
-				if (s == _NN) {
-					wrtb (0xED);
-					wrtb (0x4B + 0x10 * (r == ldDE));
+				case ldBC:
+				case ldDE:
+					s = rd_nn_nn (&ptr);
+					if (s == _NN) {
+						wrtb (0xED);
+						wrtb (0x4B + 0x10 * (r == ldDE));
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: same as above.

File: librz/asm/arch/z80/z80asm.c [hunk 6]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
+						break;
+					}
+					wrtb (0x01 + (r == ldDE) * 0x10);
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: same here.

File: librz/asm/arch/z80/z80asm.c [hunk 7]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
-			case ldHL:
-				r = rd_nn_nn (&ptr);
-				wrtb (0x21 + (r == _NN) * 9);
-				break;
-			case ldI:
-			case ldR:
-				if (!rd_a (&ptr)) {
+				case ldHL:
+					r = rd_nn_nn (&ptr);
+					wrtb (0x21 + (r == _NN) * 9);
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: same here.

File: librz/asm/arch/z80/z80asm.c [hunk 8]
```diff
@@ -1055,105 +1051,105 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
-				r = rd_sp (&ptr);
-				if (r == SPHL) {
-					wrtb (0xF9);
+				case ldI:
+				case ldR:
+					if (!rd_a (&ptr)) {
+						break;
+					}
+					wrtb (0xED);
+					wrtb (0x47 + 0x08 * (r == ldR));
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: same here.

**Review by wargio — CHANGES_REQUESTED** [review]
If i miss any of those `0x10 * (r == ldDE)` please convert them to `0x10 * (r == ldDE ? 1 : 0)`

**Comment by ghost** [comment]
@wargio all your comment are about lines I didn't edit (I only changed indentation)

But if you want, I can fix those lines.

**Comment by wargio** [comment]
> @wargio all your comment are about lines I didn't edit (I only changed indentation)
> 
> But if you want, I can fix those lines.

Yes i understand, but since you are refactoring, can you address those issues?

File: librz/asm/arch/z80/z80asm.c [hunk 9]
```diff
@@ -1055,105 +1051,106 @@ static int assemble(const char *str, unsigned char *_obuf) {
...
-				wrtb (0x43 + 0x10 * --r);
-				break;
-			case ldA:
-				if (!(r = rd_lda (&ptr))) {
+				case ld_BC:
+				case ld_DE:
+					if (!rd_a (&ptr)) {
+						break;
+					}
+					wrtb (0x02 + 0x10 * (r == ld_DE));
```
Comment (wargio) [inline librz/asm/arch/z80/z80asm.c]: i think you missed this one

**Review by wargio — CHANGES_REQUESTED** [review]
missed one. once, that one is fixed, we can merge!

**Review by wargio — APPROVED** [review]

**Pull request merged**